### PR TITLE
gh-92679: Clarify roles of asyncio protocols and transports

### PR DESCRIPTION
**Pull request by arhadthedev**
#92679

cc @1st1 @asvetlov as asyncio experts.

* Issue: gh-92679

### CODE REVIEW

**Comment by arhadthedev** [comment]
@1st1 Could you review and possibly merge please?

File: Doc/library/asyncio-protocol.rst [hunk 1]
```diff
@@ -23,21 +23,21 @@ This documentation page covers both `Transports`_ and `Protocols`_.
 
 .. rubric:: Introduction
 
-At the highest level, the transport is concerned with *how* bytes
-are transmitted, while the protocol determines *which* bytes to
-transmit (and to some extent when).
+The transport is an interface for a closable source of raw bytes,
```
Comment (gvanrossum) [inline Doc/library/asyncio-protocol.rst]: Arguably "closeable" is an ugly word and also a detail that isn't important in the first sentence of this introduction.
```suggestion
The transport is an interface for a source of raw bytes,
```

File: Doc/library/asyncio-protocol.rst [hunk 2]
```diff
@@ -23,21 +23,21 @@ This documentation page covers both `Transports`_ and `Protocols`_.
...
+while the protocol represents a consumer of these bytes that may answer back.
 
-A different way of saying the same thing: a transport is an
-abstraction for a socket (or similar I/O endpoint) while a protocol
-is an abstraction for an application, from the transport's point
-of view.
+The protocol object may either generate and consume data by its own (like a
```
Comment (gvanrossum) [inline Doc/library/asyncio-protocol.rst]:
```suggestion
The protocol object may either generate and consume data on its own (like a
```

File: Doc/library/asyncio-protocol.rst [hunk 3]
```diff
@@ -23,21 +23,21 @@ This documentation page covers both `Transports`_ and `Protocols`_.
...
+fallback background sender behind :func:`~asyncio.loop.sendfile`), be an
```
Comment (gvanrossum) [inline Doc/library/asyncio-protocol.rst:30]: I have no idea what a "fallback background sender" would be or why `sendfile` is used here. I suppose I could look it up in the source, but that's asking a bit much in a section labeled "Introduction". It should speak for itself.

File: Doc/library/asyncio-protocol.rst [hunk 4]
```diff
@@ -23,21 +23,21 @@ This documentation page covers both `Transports`_ and `Protocols`_.
...
+adapter for another I/O interface (like the
+:func:`~asyncio.StreamReader`/:func:`~asyncio.StreamWriter` backend),
+or play a role of a converter middle man between other transport and protocol
+(like the one created by :func:`~asyncio.loop.start_tls`).
 
-Yet another view is the transport and protocol interfaces
-together define an abstract interface for using network I/O and
-interprocess I/O.
+While the transport is actively puppeteered by an asyncio event loop, the
```
Comment (gvanrossum) [inline Doc/library/asyncio-protocol.rst:36]: "puppeteered" is a bit poetic for technical documentation. :-)

**Review by gvanrossum — COMMENTED** [review]
TBH I'm not convinced that this rewrite makes the matter all that much clearer to someone who doesn't already know what Protocols and Transports are.

FWIW, we didn't invent the terminology, we borrowed it from Twisted. I can never remember what they are until I see some source code -- then it falls in place easily. The second paragraph of the original, which you ruthlessly deleted, actually explains it rather clearly. You may object that it doesn't allow for a protocol/transport pair to be inserted between another transport/protocol pair, like middleware, e.g. start_tls. But the point of middleware (in this sense) is that it doesn't matter whether the "other end" is literally what you think it is (the application or the network, depending on which side you are standing), since it's always in fact some kind of abstraction. As long as it conforms to the right interface.

**Comment by kumaraditya303** [comment]
The issue https://github.com/python/cpython/issues/92679 is closed, is this still relevant? @arhadthedev

**Comment by arhadthedev** [comment]
@kumaraditya303 I believe it is. The issue had a greater scope of serious rehaul found to be worth a proper pypi library for real-world testing.

File: Doc/library/asyncio-protocol.rst (hunk 3 above)
Comment (arhadthedev) [reply Doc/library/asyncio-protocol.rst:30]: I'm sorry for postponing the stuff for so long. I've got a load of thesis work in the university so I'll be free and able to address all the comments a couple of weeks later.

**Comment by kumaraditya303** [comment]
>  I believe it is. The issue had a greater scope of serious rehaul found to be worth a proper pypi library for real-world testing.

Okay thanks! I believe Guido will review this better than me so leaving this to him.

File: Doc/library/asyncio-protocol.rst (hunk 3 above)
Comment (gvanrossum) [reply Doc/library/asyncio-protocol.rst:30]: Okay, I am patiently waiting.

**Comment by arhadthedev** [comment]
> TBH I'm not convinced that this rewrite makes the matter all that much clearer to someone who doesn't already know what Protocols and Transports are.
>
> FWIW, we didn't invent the terminology, we borrowed it from Twisted.

For a year, I've seen no evidence of the contrary so closing this PR.

**Pull request closed without merging**